### PR TITLE
feat: psr 14 implentation for event dispatcher

### PR DESCRIPTION
**Pull request by haflit21**
### Why

The PSR-14 standard for event dispatcher require that the event are send to the dispatch function. So if you using the PSR-14 standard or the `symfony/event-dispatcher >= 4.3`, this is the version to use. 

### How to test

Follow the documentation to install and configure the bundle. Then add the command in `Usage` part. After that, just run the command.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,47 +1,84 @@
...
+Note:   
+- If you are using a symfony version `>= 4.3` use the lastest version
+- For symfony versions between `2.3` and `3.0`, you can use `m6web/daemon-bundle:^1.4`
+- For PHP versions `>=5.5.9` and `<7.0` support, you can use `m6web/daemon-bundle:^3.0`
+
+*For more information about installation of plugin refers the documentation of symfony for your version.*
+     
+## Configuration  
+  
+You can optionally define events which are triggered each X iterations :
```
Comment (saiht) [inline README.md]:
```suggestion
You can optionally define events which are triggered each X iterations:  
```

File: README.md [hunk 2]
```diff
@@ -1,47 +1,84 @@
...
-m6_web_daemon:
-    iterations_events:
-        -
-            count: 10
-            name: "daemon.iteration.each.10"
-        -
-            count: 5
-            name: "daemon.iteration.each.5"
+# config/services.yaml
+seervices:
```
Comment (saiht) [inline README.md]:
```suggestion
services:
```

File: README.md [hunk 3]
```diff
@@ -1,47 +1,84 @@
...
 
-You can optionally define events which are triggered each X iterations :
+use M6Web\Bundle\DaemonBundle\Event\AbstractDaemonEvent;
 
+class EachFiveEvent extends AbstractDaemonEvent
+{
+}
+```
+
+This bundle use the PSR-14 implementation for event dispatcher so you need to refer the symfony event dispatcher in your `config/services.yaml` like this:
```
Comment (saiht) [inline README.md]:
```suggestion
This bundle use the PSR-14 implementation for event dispatcher so you need to register the symfony event dispatcher in your `config/services.yaml` like this:
```

File: src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php [hunk 4]
```diff
@@ -2,114 +2,93 @@
...
+    /** @var array */
     protected $iterationsEvents = [];
 
-    /**
-     * @var array
-     */
+    /** @var array */
     protected $iterationsIntervalCallbacks = [];
 
+    public function setEventDispatcher(EventDispatcherInterface $dispatcher = null): DaemonCommand
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php]: Couldn't this be autowired and go into the construct ? valid also for `setLoop` and `setIterationsEvents`

File: src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php [hunk 5]
```diff
@@ -2,114 +2,93 @@
...
-    /**
-     * @var LoopInterface
-     */
+    /** @var LoopInterface */
     protected $loop;
 
-    /**
-     * @var callable
-     */
+    /** @var callable */
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php]: `?callable` as it is not instantiated

File: src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php [hunk 6]
```diff
@@ -2,114 +2,93 @@
...
-     * @var EventDispatcherInterface
-     */
-    protected $dispatcher = null;
+    /** @var ?EventDispatcherInterface */
+    protected $dispatcher;
 
-    /**
-     * @var LoopInterface
-     */
+    /** @var LoopInterface */
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php]: `?LoopInterface`

File: src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php [hunk 7]
```diff
@@ -142,8 +121,6 @@ private function configureDaemonDefinition(): void
     /**
      * Define command code callback.
      *
-     * @param callable $callback
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php]: why did you remove the php doc for the param only ?

File: src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php [hunk 8]
```diff
@@ -241,48 +200,29 @@ public function daemon(InputInterface $input, OutputInterface $output)
...
      *
-     * @param string $eventName
      * @return DaemonCommand
      */
     protected function dispatchEvent(string $eventName): self
     {
-        $dispatcher = $this->getEventDispatcher();
-        if (!is_null($dispatcher)) {
-            $time = !is_null($this->startTime) ? microtime(true) - $this->startTime : 0;
+        if ($this->dispatcher !== null) {
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php]: yoda condition or remove the `!== null`

File: src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php [hunk 9]
```diff
@@ -241,48 +200,29 @@ public function daemon(InputInterface $input, OutputInterface $output)
...
+            $time = $this->startTime !== null ? microtime(true) - $this->startTime : 0;
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Command/DaemonCommand.php]: `$time = $this->startTime ? microtime(true) - $this->startTime : 0;`

File: src/M6Web/Bundle/DaemonBundle/Event/AbstractDaemonEvent.php [hunk 10]
```diff
@@ -2,91 +2,71 @@
...
+     * Get last exception class name.
      */
-    public function getCommandLastExceptionClassName()
+    public function getCommandLastExceptionClassName(): ?string
     {
         $exception = $this->command->getLastException();
 
-        if (!is_null($exception)) {
-            return get_class($exception);
+        if ($exception !== null) {
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Event/AbstractDaemonEvent.php]: yoda condition or remove `!== null`

File: src/M6Web/Bundle/DaemonBundle/Tests/Fixtures/Command/DaemonCommandConcrete.php [hunk 11]
```diff
@@ -0,0 +1,21 @@
+<?php
+
+namespace M6Web\Bundle\DaemonBundle\Tests\Fixtures\Command;
+
+use M6Web\Bundle\DaemonBundle\Command\DaemonCommand;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class DaemonCommandConcrete extends DaemonCommand
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Tests/Fixtures/Command/DaemonCommandConcrete.php]: We could remove the "concrete" and just call it `DaemonCommand` WDYT?

Comment (GautierLB) [reply src/M6Web/Bundle/DaemonBundle/Tests/Fixtures/Command/DaemonCommandConcrete.php]: Same goes for others but it's tests so as you wish

File: src/M6Web/Bundle/DaemonBundle/Tests/Fixtures/Command/DaemonCommandConcreteThrowException.php [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace M6Web\Bundle\DaemonBundle\Tests\Fixtures\Command;
+
+use M6Web\Bundle\DaemonBundle\Command\DaemonCommand;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class DaemonCommandConcreteThrowException extends DaemonCommand
+{
+    public static $exceptionMessage = null;
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Tests/Fixtures/Command/DaemonCommandConcreteThrowException.php]: 😱 maybe a setter ?

File: composer.json [hunk 13]
```diff
@@ -2,31 +2,34 @@
...
+        "phpunit/phpunit": "^8.4"
     },
     "autoload" : {
-        "psr-0" : {
-            "" : "src/"
-        }
+        "psr-0" : { "" : "src/"}
+    },
+    "conflict": {
+        "symfony/event-dispatcher": "<4.3"
```
Comment (f-dumas) [inline composer.json]: :+1:

File: src/M6Web/Bundle/DaemonBundle/Tests/Fixtures/Command/DaemonCommandConcreteThrowStopException.php [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+class DaemonCommandConcreteThrowStopException extends DaemonCommand
+{
+    /** @var int */
+    public const MAX_ITERATION = 7;
+
+    /** @var string */
+    public const EXCEPTION_MESSAGE = 'Stop loop exception';
+
+    /** @var int */
+    public static $count = 0;
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Tests/Fixtures/Command/DaemonCommandConcreteThrowStopException.php]: _sad encapsulation noise_ 😢

File: src/M6Web/Bundle/DaemonBundle/Tests/Unit/DependencyInjection/M6WebDaemonExtensionTest.php [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+        $this->assertEquals('test', $iterationEvent['name']);
+    }
+
+    /**
+     * @throws \Exception
+     */
+    public function testMissingCountConfig(): void
+    {
+        $container = $this->getContainerForConfiguration('bad-parameters-config');
+        $this->expectException(InvalidTypeException::class);
```
Comment (GautierLB) [inline src/M6Web/Bundle/DaemonBundle/Tests/Unit/DependencyInjection/M6WebDaemonExtensionTest.php]: This could be in the phpdochttps://phpunit.readthedocs.io/fr/latest/writing-tests-for-phpunit.html#writing-tests-for-phpunit-exceptions-examples-exceptiontest-php

File: composer.json [hunk 16]
```diff
@@ -2,31 +2,34 @@
...
     "require" : {
-        "php" : ">=7.1",
+        "php" : ">=7.2",
         "ext-pcntl" : "*",
-        "symfony/console" : "~3.0 || ~4.0",
-        "symfony/framework-bundle" : "~3.0 || ~4.0",
-        "symfony/yaml": "~3.0 || ~4.0",
+        "psr/event-dispatcher": "^1.0",
+        "symfony/console" : "~4.0",
+        "symfony/framework-bundle" : "~4.0",
```
[the remark this answers is not in the document]
Comment (haflit21) [reply composer.json]: That's take the >=4.0 version but i can force the >=4.3. It will be better in fact
